### PR TITLE
Smoketests

### LINKED ISSUE #451: CI Smoke tests
As there is a need to validate the features we've added I would suggest we add something similar to this setup:

```
FROM k0sproject/k0s:latest
RUN apk add openssl
RUN curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 \
    && chmod +x get_helm.sh && ./get_helm.sh
```
```
docker build . -t k0s1
docker run -d --name k0s --hostname k0s --privileged -v /var/lib/k0s -p 6445:6443 k0s1

docker exec k0s helm repo add kubeshop "https://kubeshop.github.io/helm-charts/"
docker exec k0s helm repo update
docker exec k0s helm install kusk-gateway kubeshop/kusk-gateway -n kusk-system --create-namespace
docker exec k0s kubectl get pods -n kusk-system

kubectl apply -f api.yaml #There isn't any at the moment

# do curl magic here
```

### PR DESCRIPTION
**Pull request by jasmingacic**
This is the initial run at smoke tests. 
Ideally we would run these on a GKE worker cluster.

Running for example `make basic` will trigger following:
- spin up local image registry
- build local image 
- deploy `make deploy` with change to kustomization.yaml adding a flag to replace (clumsy hack it needs serious improvement
- deploys recent build to the accessible cluster
- hits smoketests/basic (in case `make testsomething` was invoked if testsomething is contained in Makefile.variables then github.com/kubeshop/kusk-gateway/smoketests/testsomething will be tested ) just a convenience if we decide to have more complex tests
- after finishing it will remove everything (it won't clean up if something fails needs to be improved) 

This is important so we can release stuff easily without worrying about breaking previous functionalities. 
Also it needs more scenarios like curl calls to newly deployed endpoints to confirm caching, rate_limiting, auth and of course if the deployed API is sane and many more) 

The tests need to be executed whenever a PR is submitted and it should run on GH custom workers (@ypoplavs)

Closes #451

### CODE REVIEW

**Comment by jasmingacic** [comment]
For whichever reason setup-envtest doesn't get fetched  so I skipped over v1alpha1 and controllers tests.

**Comment by jasmingacic** [comment]
To run the tests do the following:
1. checkout the branch `git checkout smoketests` 
2. run make check-basic for basic tests or make check-mocking for mock tests
3. if a test fails make sure you don't commit kustomization.yam like I did (still need to figure this one out)

File: smoketests/common/commonSuite.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	s.NoError(err)
+
+	deploy := apps.Deployment{}
+	counter := 0
+	for {
+		if counter == 100 {
+			break
+		}
+		s.NoError(s.Cli.Get(context.Background(), client.ObjectKey{Namespace: KuskNamespace, Name: KuskManager}, &deploy))
+		// if counter == 0 {
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply smoketests/common/commonSuite.go]: Leftovers which i removed.

Thanks

File: smoketests/common/commonSuite.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		// 	time.Sleep(5 * time.Second)
+		// }
+		fmt.Println(" deploy.Status.AvailableReplicas", deploy.Status.AvailableReplicas)
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply smoketests/common/commonSuite.go]: totally unnecessary i removed it

File: Makefile [hunk 3]
```diff
@@ -237,3 +239,20 @@ curl -LO $${PB_REL}/download/v$${VERSION}/$${FILENAME} ; unzip $${FILENAME} -d $
...
 }
 endef
+
+start-smoke-tests: start-smoke-tests deploy-local-registry deploy-envoyfleet
+	@docker rm smoke-registry -f
+	@docker run -d -p 50000:5000 --name smoke-registry registry:2
+	@docker build -t localhost:50000/kusk-gateway:smoke -f ./build/manager/Dockerfile .
+	@docker push localhost:50000/kusk-gateway:smoke
+	
+deploy-local-registry: envtest
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply Makefile:249]: kustomize only

File: Makefile [hunk 4]
```diff
@@ -237,3 +239,20 @@ curl -LO $${PB_REL}/download/v$${VERSION}/$${FILENAME} ; unzip $${FILENAME} -d $
...
+	echo $(shell pwd)
+	bin/kustomize build smoketests/common  | kubectl apply -f -
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply Makefile:251]: can you give it a try and push changes to this branch

File: Makefile [hunk 5]
```diff
@@ -237,3 +239,20 @@ curl -LO $${PB_REL}/download/v$${VERSION}/$${FILENAME} ; unzip $${FILENAME} -d $
 rm -rf $${TMP_DIR} ;\
 }
 endef
+
+start-smoke-tests: start-smoke-tests deploy-local-registry deploy-envoyfleet
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply Makefile:243]: copy paste mistake

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasmingacic) [reply Makefile:251]: if i change to that i get a different error:
```
bash: /root/go/src/github.com/kubeshop/kusk-gateway/smoketests/bin/kustomize: No such file or directory
```

somehow it looks for binaries in smoketests

File: smoketests/mocking/mocking_test.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+	defer resp.Body.Close()
+
+	m.Equal(200, resp.StatusCode)
+
+	o, _ := io.ReadAll(resp.Body)
+	res := map[string]string{}
+	m.NoError(json.Unmarshal(o, &res))
+
+	m.Equal(res["message"], "Hello from a mocked response!")
```
[the remark this answers is not in the document]
Comment (kylehodgetts) [reply smoketests/mocking/mocking_test.go:87]: Yes, this is correct

**Comment by jasmingacic** [comment]
Folks let's have this merged as we need it to test our upcoming release
Can you just approve it

**Review by kylehodgetts — APPROVED** [review]

**Comment by jasmingacic** [comment]
@kylehodgetts thank you sir
@mbana give me some approval :D

File: smoketests/Makefile [hunk 7]
```diff
@@ -0,0 +1,14 @@
+include Makefile.variables
+
+
+$(smoketests): #sandbox
+	go mod tidy
+	go test -count=1 -v github.com/kubeshop/kusk-gateway/smoketests/$(subst check-,,$@)
+	# kubectl delete -f samples/hello-world/deployment.yaml
+
+
+sandbox:
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply smoketests/Makefile:10]: I commented it out as we don't need it for mock tests but we will need it for "non-mock" tests

File: smoketests/basic/api.yaml [hunk 8]
```diff
@@ -0,0 +1,315 @@
+apiVersion: gateway.kusk.io/v1alpha1
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply smoketests/basic/api.yaml:1]: It doesn't really matter we can tweak this as we go

File: smoketests/mocking/envoyfleet.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
+apiVersion: gateway.kusk.io/v1alpha1
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply smoketests/mocking/envoyfleet.yaml:1]: yes

File: Makefile [hunk 10]
```diff
@@ -157,7 +159,7 @@ uninstall: manifests $(KUSTOMIZE) ## Uninstall CRDs from the K8s cluster specifi
 
 .PHONY: deploy
 deploy: manifests $(KUSTOMIZE) ## Deploy controller to the K8s cluster specified in ~/.kube/config.
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
+	$(KUSTOMIZE) build config/default  | kubectl apply -f -
```
Comment (kylehodgetts) [inline Makefile:162]: added extra space?

File: Makefile [hunk 11]
```diff
@@ -237,3 +239,20 @@ curl -LO $${PB_REL}/download/v$${VERSION}/$${FILENAME} ; unzip $${FILENAME} -d $
...
+deploy-local-registry: $(ENVTEST) $(KUSTOMIZE)
+	echo $(shell pwd)
+	bin/kustomize build smoketests/common  | kubectl apply -f -
+
+.PHONY: $(smoketests)
+$(smoketests): start-smoke-tests
+	$(MAKE) -C smoketests $@
+	@rm -rf smoketests/bin
+
+check-all: check-basic check-mocking
```
Comment (kylehodgetts) [inline Makefile:258]: check-all in the root Makefile doesn't really tell us anything. Consider renaming it to something related to smoke tests. Also make sure that the help is displayed correctly when running `make help` and that it includes this option

Comment (kylehodgetts) [reply Makefile:258]: We should add another section for testing. Currently `make help` returns this

```
❯ make help

Usage:
  make <target>

General
  help             Display this help.

Development
  create-env       Spin up a local development cluster with Minikube and install kusk Gateway
  deploy-envoyfleet  Deploy k8s resources for the single Envoy Fleet
  delete-env       Destroy the local development Minikube cluster
  manifests        Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
  generate         Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
  fmt              Run go fmt against code.
  vet              Run go vet against code.
  test             Run tests.
  testing          Run the integration tests from development/testing and then delete testing artifacts if succesfull.
  goproxy          Starts local goproxy docker instance for the faster builds. Make sure to set your shell environment variable (e.g. put into .bashrc "export GOPROXY=http://<your_ip_address>:8085".
  docker-images-cache  Saves locally frequently used container images and uploads them to Minikube to speed up the development.

Build
  build            Build manager binary.
  run              Run a controller from your host, proxying it inside the cluster.
  docker-build-manager  Build docker image with the manager.
  docker-build     Build docker images for all apps
  docker-build-manager-debug  Build docker image with the manager and debugger.

Deployment
  install-local    Install CRDs and Envoy into the K8s cluster specified in ~/.kube/config.
  uninstall-local  Uninstall CRDs and Envoy from the K8s cluster specified in ~/.kube/config.
  install          Install CRDs into the K8s cluster specified in ~/.kube/config.
  uninstall        Uninstall CRDs from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
  deploy           Deploy controller to the K8s cluster specified in ~/.kube/config.
  deploy-debug     Deploy controller with debugger to the K8s cluster specified in ~/.kube/config.
  undeploy         Undeploy controller from the K8s cluster specified in ~/.kube/config.
  update           Runs deploy, docker build and restarts kusk-gateway-manager deployment to pick up the change
  update-debug     Runs Debug configuration deploy, docker build and restarts kusk-gateway-manager deployment to pick up the change
  cycle            Triggers kusk-gateway-manager deployment rollout restart to pick up the new container image with the same tag
  cycle-envoy      Triggers all Envoy pods in the cluster to restart
  envtest          Download envtest-setup locally if necessary.
  tools            Install all tools
```

File: smoketests/Makefile [hunk 12]
```diff
@@ -0,0 +1,14 @@
+include Makefile.variables
+
+
+$(smoketests): 
+	go mod tidy
+	go test -count=1 -v github.com/kubeshop/kusk-gateway/smoketests/$(subst check-,,$@)
+	# kubectl delete -f samples/hello-world/deployment.yaml
```
Comment (kylehodgetts) [inline smoketests/Makefile:7]: Should remove the commented out command if not needed

File: smoketests/common/commonSuite.go [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	corev1.AddToScheme(scheme)
+	apps.AddToScheme(scheme)
+
+	s.Cli, err = client.New(config, client.Options{Scheme: scheme})
+	s.NoError(err)
+
+	deploy := apps.Deployment{}
+	counter := 0
+	for {
+		if counter == 100 {
```
Comment (kylehodgetts) [inline smoketests/common/commonSuite.go:45]: Then you can remove this

File: Makefile (hunk 11 above)
Comment (jasmingacic) [reply Makefile:258]: It won't work as I'm pulling names of smoketest from smoketests/Makefile.variables. 
Probably I don't know how to do it.

Smoketests are for our internal usage anyways.

File: smoketests/common/commonSuite.go [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+			break
+		}
+		s.NoError(s.Cli.Get(context.Background(), client.ObjectKey{Namespace: KuskNamespace, Name: KuskManager}, &deploy))
+
+		if deploy.Status.AvailableReplicas > 0 && deploy.Status.ReadyReplicas > 0 {
+			break
+		} else {
```
Comment (kylehodgetts) [inline smoketests/common/commonSuite.go:52]: Don't need else if we have a break in the if branch

File: smoketests/common/commonSuite.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+	kuskv1.AddToScheme(scheme)
+	corev1.AddToScheme(scheme)
+	apps.AddToScheme(scheme)
+
+	s.Cli, err = client.New(config, client.Options{Scheme: scheme})
+	s.NoError(err)
+
+	deploy := apps.Deployment{}
+	counter := 0
+	for {
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply smoketests/common/commonSuite.go:44]: totally missed it :D

**Pull request merged**